### PR TITLE
Allow uploading revisions when an editor hasn't been assigned

### PR DESCRIPTION
**Pull request by plourenco**
Once an editable has been submitted, it is currently impossible to upload a new version of it. The improvement would allow a new revision to be made at any time (but disallowed when an editor has been assigned - to prevent him from seeing an old version).

### CODE REVIEW

File: indico/modules/events/editing/controllers/backend/timeline.py [hunk 1]
```diff
@@ -50,17 +50,19 @@ def get_file_context(self):
...
-        if last_rev:
-            found = next((f for f in last_rev.files if f.id == id), None)
-            if found:
-                with found.open() as stream:
-                    return self._save_file(found, stream)
+        files = []
+        if self.contrib.editables:
+            files = [file.file for e in self.contrib.editables for file in e.revisions[-1].files]
+        if self.contrib.paper and (last_rev := self.contrib.paper.get_last_revision()):
+            files += [f for f in last_rev.files]
```
[the remark this answers is not in the document]
Comment (plourenco) [reply indico/modules/events/editing/controllers/backend/timeline.py:62]: Yes, that looks stupid

File: indico/modules/events/editing/controllers/backend/timeline.py [hunk 2]
```diff
@@ -50,17 +50,20 @@ def get_file_context(self):
...
-            if found:
-                with found.open() as stream:
-                    return self._save_file(found, stream)
+        files = []
+        if self.contrib.editables:
+            files = [file.file for e in self.contrib.editables
+                     if e.type == self.editable.type for file in e.revisions[-1].files]
+        if self.contrib.paper and (last_rev := self.contrib.paper.get_last_revision()):
+            files.extend(last_rev.files)
+        if found := next((f for f in files if f.id == id), None):
```
Comment (ThiefMaster) [inline indico/modules/events/editing/controllers/backend/timeline.py:64]: `EditingRevisionFile` and `PaperFile` are in different tables, so it's possible for an ID collision between them -> as discussed on Mattermost we need e.g. a query param indicating the source of the file

Comment (plourenco) [reply indico/modules/events/editing/controllers/backend/timeline.py]: It was a bit more complicated than foreseen, but we're past it.

File: indico/modules/events/editing/controllers/backend/timeline.py [hunk 3]
```diff
@@ -50,17 +50,23 @@ def get_file_context(self):
...
-                    return self._save_file(found, stream)
+    def _process(self, id, paper_id):
+        if bool(id) == bool(paper_id):
+            raise BadRequest
+        files = []
+        if id and self.contrib.editables:
+            files = [file.file for e in self.contrib.editables
+                     if e.type == self.editable.type for file in e.revisions[-1].files]
+        if paper_id and self.contrib.paper and (last_rev := self.contrib.paper.get_last_revision()):
+            files.extend(last_rev.files)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/controllers/backend/timeline.py:66]: Why not `files = ...` since we never expect to get files from multiple sources?

Comment (plourenco) [reply indico/modules/events/editing/controllers/backend/timeline.py:66]: Initially, I was expecting to have both but didn't update accordingly.

**Review by ThiefMaster — APPROVED** [review]

**Review by ThiefMaster — APPROVED** [review]

**Pull request merged**